extensions/standard-processors/tests/unit/PutUDPTests.cpp

```diff
@@ -0,0 +1,112 @@
+    if (bind_result == utils::net::SocketError) {
+      throw std::runtime_error{utils::StringUtils::join_pack("bind: ", utils::net::get_last_socket_error().message())};
+    }
+  }
+
+  struct ReceiveResult {
+    std::string remote_address;
+    std::string message;
+  };
+
+  [[nodiscard]] ReceiveResult receive(const size_t max_message_size = 8192) const {
+    ReceiveResult result;
+    result.message.resize(max_message_size);
+    sockaddr_storage remote_address{};
+    socklen_t addrlen = sizeof(remote_address);
+    const auto recv_result = recvfrom(open_socket_.socket_.get(), result.message.data(), result.message.size(), 0, std::launder(reinterpret_cast<sockaddr*>(&remote_address)), &addrlen);
+    if (recv_result == utils::net::SocketError) {
+      throw std::runtime_error{utils::StringUtils::join_pack("recvfrom: ", utils::net::get_last_socket_error().message())};
+    }
+    result.message.resize(gsl::narrow<size_t>(recv_result));
+    result.remote_address = utils::net::sockaddr_ntop(std::launder(reinterpret_cast<sockaddr*>(&remote_address)));
+    return result;
+  }
+
+  std::unique_ptr<addrinfo, utils::net::addrinfo_deleter> resolved_names_;
+  utils::net::OpenSocketResult open_socket_;
+};
+}  // namespace
+
+// Testing the failure relationship is not required, because since UDP in general without guarantees, flow files are always routed to success, unless there is
+// some weird IO error with the content repo.
+TEST_CASE("PutUDP", "[putudp]") {
+  const auto putudp = std::make_shared<PutUDP>("PutUDP");
+  auto random_engine = std::mt19937{std::random_device{}()};  // NOLINT: "Missing space before {  [whitespace/braces] [5]"
+  // most systems use ports 32768 - 65535 as ephemeral ports, so avoid binding to those
+  const auto port = std::uniform_int_distribution<uint16_t>{10000, 32768 - 1}(random_engine);
+  const auto port_str = std::to_string(port);
+
+  test::SingleInputTestController controller{putudp};
+  LogTestController::getInstance().setTrace<PutUDP>();
+  LogTestController::getInstance().setTrace<core::ProcessContext>();
+  LogTestController::getInstance().setLevelByClassName(spdlog::level::trace, "org::apache::nifi::minifi::core::ProcessContextExpr");
+  putudp->setProperty(PutUDP::Hostname, "${literal('localhost')}");
+  putudp->setProperty(PutUDP::Port, utils::StringUtils::join_pack("${literal('", port_str, "')}"));
+
+  DatagramListener listener{"localhost", port_str.c_str()};
+
+  {
+    const char* const message = "first message: hello";
+    const auto result = controller.trigger(message);
+    const auto& success_flow_files = result.at(PutUDP::Success);
+    REQUIRE(success_flow_files.size() == 1);
+    REQUIRE(result.at(PutUDP::Failure).empty());
+    REQUIRE(controller.plan->getContent(success_flow_files[0]) == message);
+    auto receive_result = listener.receive();
+    REQUIRE(receive_result.message == message);
+    REQUIRE(!receive_result.remote_address.empty());
+  }
+
+  {
```

Comment:
I want to reuse the listener so that the test only needs one port from the host system (+ one for the processor), not two.

Reply:
Once `UniqueSocketHandle` has a destructor, `DatagramListener` will close the socket on destruction, so we will still only have two sockets open at the same time.  Does that take a very long, or is there another reason for not wanting to open a port 3 times instead of 2?

Reply:
I just wanted to minimize the chance of collision and avoid SO_REUSEADDR, because [it can cause problems](https://stackoverflow.com/a/3233022). And in general it's not nice that unit tests open ports, but it's hard to avoid here.